### DISCUSSION TITLE
How's my eglot-rename-file function?

### DISCUSSION THREAD

**Discussion by timcharper** [joaotavora/eglot#1396] [Show and tell]
This is my first time working with eglot internals so I'm probably abusing things and doing it wrong. I'm pretty inexperienced with lsp in general.

Anyways, I really wanted typescript rename-files support in eglot so I had one less reason to open vs-code. Turns out the lsp server supports the [WillRenameFiles](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#workspace_willRenameFiles) command.

Here's my code:

```elisp
(defun eglot-rename-file ()
  (interactive)
  (let* ((new-name (read-string "New name: "))
         (old-name (buffer-file-name))
         (old-uri (eglot-path-to-uri old-name))
         (new-uri (eglot-path-to-uri new-name)))
    ;; First, send the LSP command
    (let ((edits (eglot--request (eglot--current-server-or-lose)
                                 :workspace/willRenameFiles
                                 `(:files [(:oldUri ,old-uri :newUri ,new-uri)]))))
      (eglot--apply-workspace-edit edits :workspace/willRenameFiles))
    ;; ;; Check if the LSP operation was successful; here we assume it was since `eglot-execute` returns nil on success.
    ;; ;; Now, perform the actual file renaming
    (when (file-exists-p old-name)
      (rename-file old-name new-name 1))
    ;; Update the buffer to reflect the new file name
    (set-visited-file-name new-name)

    ;; force redisplay of buffer / modeline
    (set-buffer-modified-p nil)

    (message "File renamed from %s to %s" old-name new-name)

    ;; manually send that we closed the old file, since the buffer file-name was changed without triggering the usual close/open events
    (jsonrpc-notify (eglot--current-server-or-lose)
                    :textDocument/didClose
                    `(:textDocument (:uri ,old-uri)))
    
    (jsonrpc-notify (eglot--current-server-or-lose)
                    :workspace/didRenameFiles
                    `(:files [(:oldUri ,old-uri :newUri ,new-uri)]))))
```

Please roast it.

**Comment by timcharper** [comment]
I don't know if didRenameFiles implies didClose

**Comment by timcharper** [comment]
So, using this for a while, there's some issues with it. Might be obvious to @joaotavora ?

One: I have to revert the file in order for the lsp server to recognize that the new file is open. I'm not sure why this is, since I send the rename commands, and, the "didOpen" event is sent automatically for the new file.

Two: renames when a path is changed don't update the file itself (i.e. fix typescript relative imports), like happens in vs-code.

But, overall, this has been super useful. I'll update it as I have fixes. Perhaps one day we can get this functionality in to eglot.

**Reply by joaotavora** [reply]
Thanks for your work.  It would be useful to have an annotated transcript of LSP/JSONRPC communication from the time that you M-x `eglot-rename-file`.

Also `rename-visited-file` sounds relevant for the Emacs side of things.

Also it would be nice to gather which servers support this apart from (what I presume is) the typescript server.  Does clangd or rust-analyzer support this?

Also it would be nice to have tests and to have the whole contribution presented in bug-gnu-emacs@gnu.org as a patch to Emacs master's `lisp/progmodes/eglot.el` and `test/lisp/progmodes/eglot-tests.el`
